include/fmt/ranges.h

```diff
@@ -357,42 +357,56 @@ template <typename Char, typename... T>
 struct formatter<tuple_join_view<Char, T...>, Char> {
   template <typename ParseContext>
   FMT_CONSTEXPR auto parse(ParseContext& ctx) -> decltype(ctx.begin()) {
-    return ctx.begin();
+    return parse_subroutine(ctx, index_helper<sizeof...(T)>{});
   }
 
   template <typename FormatContext>
-  auto format(const tuple_join_view<Char, T...>& value, FormatContext& ctx) ->
-      typename FormatContext::iterator {
-    return format(value, ctx, detail::make_index_sequence<sizeof...(T)>{});
+  auto format(const tuple_join_view<Char, T...>& value,
+              FormatContext& ctx) const -> typename FormatContext::iterator {
+    return format_subroutine(value, ctx, index_helper<sizeof...(T)>{});
   }
 
  private:
-  template <typename FormatContext, size_t... N>
-  auto format(const tuple_join_view<Char, T...>& value, FormatContext& ctx,
-              detail::index_sequence<N...>) ->
-      typename FormatContext::iterator {
-    using std::get;
-    return format_args(value, ctx, get<N>(value.tuple)...);
+  std::tuple<formatter<typename std::decay<T>::type, Char>...> formatters;
+
+  template <size_t> struct index_helper {};
+
+  template <typename ParseContext>
+  FMT_CONSTEXPR auto parse_subroutine(ParseContext& ctx, index_helper<0>)
+      -> decltype(ctx.begin()) {
+    return ctx.begin();
+  }
+
+  template <typename ParseContext, size_t N>
+  FMT_CONSTEXPR auto parse_subroutine(ParseContext& ctx, index_helper<N>)
+      -> decltype(ctx.begin()) {
+    auto end = std::get<sizeof...(T) - N>(formatters).parse(ctx);
+    if (N > 1) {
+      auto end1 = parse_subroutine(ctx, index_helper<N - 1>{});
+      if (end != end1)
+        throw fmt::format_error{
+            "specs have different end for different types in tuple"};
+    }
+    return end;
   }
 
   template <typename FormatContext>
-  auto format_args(const tuple_join_view<Char, T...>&, FormatContext& ctx) ->
+  auto format_subroutine(const tuple_join_view<Char, T...>&, FormatContext& ctx,
+                         index_helper<0>) const ->
       typename FormatContext::iterator {
-    // NOTE: for compilers that support C++17, this empty function instantiation
```

Comment:
This function is reserved for empty tuple, cannot remove and will not be entered if `sizeof...(T)>0`.